File: src/dsl/command.rs
use crate::object::rect::{BorderStyle, ContentAlign, ContentOverflow, Side};
use crate::object::DrawObject;

/// Represents a parsed DSL command.
#[derive(Debug)]
pub enum DslCommand {
    /// Canvas size declaration with optional border and global defaults.
    Canvas {
        width: CanvasSize,
        height: CanvasSize,
        border: Option<BorderStyle>,
        content_overflow: Option<ContentOverflow>,
        content_align: Option<ContentAlign>,
    },
    /// Collision mode declaration.
    Collision(bool),
    /// A drawable object (rect, text, hline, vline — NOT arrow).
    Object(DrawObject),
    /// An unresolved arrow referencing object IDs (resolved later).
    Arrow {
        src_id: String,
        src_side: Side,
        dst_id: String,
        dst_side: Side,
        line: usize,
    },
}

/// Canvas dimension: explicit size or auto.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CanvasSize {
    Fixed(usize),
    Auto,
}

File: internal/dsl/command.go
package dsl

import "github.com/silee9019/unicode-diagram/internal/object"

type DslCommand interface {
	isDslCommand()
}

type CanvasSize struct {
	IsAuto bool
	Value  int
}

type CanvasCmd struct {
	Width, Height   CanvasSize
	Border          *object.BorderStyle
	ContentOverflow *object.ContentOverflow
	ContentAlign    *object.ContentAlign
}

func (c *CanvasCmd) isDslCommand() {}

type CollisionCmd struct {
	On bool
}

func (c *CollisionCmd) isDslCommand() {}

type ObjectCmd struct {
	Object object.DrawObject
}

func (c *ObjectCmd) isDslCommand() {}

type ArrowCmd struct {
	SrcID, DstID     string
	SrcSide, DstSide object.Side
	Head             rune
	HasHead          bool
	Both             bool
	Legend           *object.Legend
	Line             int
}

func (c *ArrowCmd) isDslCommand() {}

type ArrowheadCmd struct {
	Ch rune
}

func (c *ArrowheadCmd) isDslCommand() {}
